service/service.go

```diff
@@ -52,17 +70,64 @@ func (svc *Service) Start() {
 
 				message.Commit()
 				log.Debug("message committed", log.Data{"instance_id": instanceID})
-			case errorMessage := <-svc.Consumer.Errors():
-				log.Error(fmt.Errorf("aborting"), log.Data{"message_received": errorMessage})
-				svc.Consumer.Closer() <- true
-				svc.Producer.Closer() <- true
-				exitCh <- true
-				return
-			case <-healthCh:
+			case consumerError := <-svc.Consumer.Errors():
+				log.Error(fmt.Errorf("aborting consumer"), log.Data{"message_received": consumerError})
+				close(serviceLoopDone)
+			case producerError := <-svc.Producer.Errors():
+				log.Error(fmt.Errorf("aborting producer"), log.Data{"message_received": producerError})
+				close(serviceLoopDone)
+			case <-apiErrors:
+				log.Error(fmt.Errorf("server error forcing shutdown"), nil)
+				close(serviceLoopDone)
 			}
 		}
 	}()
-	<-exitCh
 
-	log.Info("service dimension extractor stopped", nil)
+	select {
+	case <-serviceLoopDone:
+		log.Debug("Quitting after done was closed", nil)
+	case signal := <-signals:
+		log.Debug("Quitting after os signal received", log.Data{"signal": signal})
+	}
+
+	childContext, childCancel := context.WithTimeout(ctx, svc.Shutdown)
+
+	waitGroup := int32(0)
+
+	atomic.AddInt32(&waitGroup, 1)
+	go func() {
+		log.Debug("Attempting to stop listening to kafka consumer group", nil)
+		svc.Consumer.StopListeningToConsumer(childContext)
+		log.Debug("Successfully stopped listening to kafka consumer group", nil)
+		listeningToConsumerGroupStopped <- true
+		<-readyToCloseOutboundConnections
+		log.Debug("Attempting to close http server", nil)
+		api.Close(childContext)
```

Comment:
this function returns an error which isnt being checked here - in all of these log statements, can we really say 'successfully' until we've checked any available errors?